csrc/cache_kernels.cu

```diff
@@ -186,6 +188,54 @@ __global__ void reshape_and_cache_kernel(
   }
 }
 
+template<typename attn_dtype, typename cache_dtype>  // cache_dtype can only be int8_t for now
+__global__ void reshape_and_cache_quantized_kernel(
+  const attn_dtype* __restrict__ key,     // [num_tokens, num_heads, head_size]
+  const attn_dtype* __restrict__ value,   // [num_tokens, num_heads, head_size]
+  cache_dtype* __restrict__ key_cache,     // [num_blocks, num_heads, head_size/x, block_size, x]
+  cache_dtype* __restrict__ value_cache,   // [num_blocks, num_heads, head_size, block_size]
+  const int* __restrict__ slot_mapping, // [num_tokens]
+  const int key_stride,
+  const int value_stride,
+  const int num_heads,
+  const int head_size,
+  const int block_size,
+  const int x,
+  const float k_scale,
+  const float k_zp,
+  const float v_scale,
+  const float v_zp) {
+  const int token_idx = blockIdx.x;
+  const int slot_idx = slot_mapping[token_idx];
+  const int block_idx = slot_idx / block_size;
+  const int block_offset = slot_idx % block_size;
+
+  const int n = num_heads * head_size;
+  for (int i = threadIdx.x; i < n; i += blockDim.x) {
+    const int src_key_idx = token_idx * key_stride + i;
+    const int src_value_idx = token_idx * value_stride + i;
+
+    const int head_idx = i / head_size;
+    const int head_offset = i % head_size;
+    const int x_idx = head_offset / x;
+    const int x_offset = head_offset % x;
+
+    const int tgt_key_idx = block_idx * num_heads * (head_size / x) * block_size * x
+                            + head_idx * (head_size / x) * block_size * x
+                            + x_idx * block_size * x
+                            + block_offset * x
+                            + x_offset;
+    const int tgt_value_idx = block_idx * num_heads * head_size * block_size
+                              + head_idx * head_size * block_size
+                              + head_offset * block_size
+                              + block_offset;
+    // TODO (Lin Pengyun): use vector reading and quantization to improve IO ultilization
+    attn_dtype tgt_key = __ldg(&key[src_key_idx]);
+    key_cache[tgt_key_idx] = quant(tgt_key, k_scale, k_zp);
+    attn_dtype tgt_value = __ldg(&value[src_value_idx]);
+    value_cache[tgt_value_idx] = quant(tgt_value, v_scale, v_zp);
+  }
+}
```

Comment:
There are lot of redundant codes in `reshape_and_cache_quantized_kernel` compared `reshape_and_cache__kernel`. Is it better to merge to one function?

Reply:
We have merged the functions and eliminated the redundant code.